### ISSUE TITLE
Greedy decoding is not reproducible: MTP (3/5 prompts) and the CUDA expert tier (2/5) both change the output

### ISSUE THREAD

**Issue by dnnspaul** [JustVugg/colibri#100]
The README states that MTP speculative decoding is **"Lossless — *and stays lossless under sampling* via rejection sampling."** On my machine it is not: with `--temp 0` (greedy), enabling MTP changes the generated text on 3 of 5 test prompts. The CUDA expert tier independently changes it on 2 of 5.

The engine itself is perfectly deterministic — repeated runs of the same config are byte-identical — so this is not run-to-run noise.

## Result

Reference = CPU path, `MTP=0`. Same prompt, `--temp 0 --ngen 32`, `RAM_GB=65`, everything else held constant.

| prompt | `MTP=1` (CPU) | CUDA expert tier (`MTP=0`) |
|---|---|---|
| capital of Germany + landmark | **diverges** | match |
| why the sky appears blue | **diverges** | **diverges** |
| first five primes | match | match |
| what a hummingbird eats | **diverges** | **diverges** |
| 17 × 23 | match | match |
| | **3/5** | **2/5** |

Most divergences are benign — an equally valid continuation. But not all. Prompt 1 produces **degenerate output** with MTP on:

```
MTP=0 (reference):  ...a major cultural and political hub renowned for its rich and complex
                    history. One of its most famous landmarks is the Brandenburg Gate
MTP=1:              ...a major cultural and political hub renowned for for its rich history
                                                                     ^^^^^^^ duplicated token
```

A benign example (prompt 4), where all three answers are correct but the token stream forks:

```
reference : ...scatter sunlight more efficiently at shorter wavelengths. Because blue light
MTP=1     : ...scatter sunlight more efficiently at shorter wavelengths. As sunlight enters
CUDA tier : ...scatter sunlight more efficiently at shorter wavelengths. As sunlight travels
```

## Ruling out the learning cache

`.coli_usage` grows after every run, so a naive A/B is confounded — later runs have a larger pin. To control for this, each prompt was run **reference → MTP → CUDA → reference again**, with the reference last:

- The trailing reference was **byte-identical to the first on 5/5 prompts.**
- Pin size grew across the sequence (757 → 1288 experts) with no effect on the reference output.

So neither cache state nor pin growth explains the divergence. MTP and the CUDA tier are the cause.

## Reproduce

```bash
P="In one short paragraph: what is the capital of Germany, and name one famous landmark there?"

COLI_CUDA=0 RAM_GB=65 MTP=0 ./coli run --ngen 32 --temp 0 "$P"   # clean
COLI_CUDA=0 RAM_GB=65 MTP=1 ./coli run --ngen 32 --temp 0 "$P"   # "renowned for for"

COLI_CUDA=1 COLI_GPU=0 CUDA_EXPERT_GB=30 RAM_GB=65 MTP=0 \
  ./coli run --ngen 32 --temp 0 "$P"                             # diverges on other prompts
```

## Notes / speculation

Since divergences cluster on near-tie tokens (sentence boundaries, connectives), the likely story is that MTP's verification path and the CUDA kernels each produce logits that differ from the CPU path in the last bits, flipping tokens whose top-2 are nearly equal. That is expected for a *different backend*, but MTP is supposed to be exactly equivalent to greedy — a verified draft token should be accepted only if it matches what the main model would have emitted. The duplicated-token case suggests the acceptance/rollback bookkeeping, not just float noise.

The MTP head here is the **int8** one (not the int4 head from #8): acceptance is healthy at 38–46%, 2.1–2.5 tokens/forward, so speculation is genuinely engaging.

## Environment

- colibrì `f38e74e`, gcc 14.2.0, `ARCH=native` (AVX-512 VNNI), `CUDA=1` (CUDA 13.0, `sm_120`)
- AMD Ryzen 7 9800X3D (8C/16T) · 94 GB RAM (70 GB in WSL2) · RTX 5090 32 GB
- Samsung 9100 PRO PCIe 5.0, ext4 inside a WSL2 VHDX · Debian, kernel 6.6.87.2
- Model: `mateogrgic/GLM-5.2-colibri-int4-with-int8-mtp` (int8 MTP heads: 3527131672 / 5366238584 / 1065950496)

**Comment by JustVugg** [comment]
This is an exceptionally well-controlled report — ruling out the learning cache with the reference-last A/B (byte-identical trailing reference on 5/5 despite the pin growing 757→1288) is exactly the right control, and it isolates the cause cleanly. You've found a real gap between the claim and the reality, and you're right. Here's what's actually happening, and it's **not a logic bug** — it's numerical:

**MTP verification is lossless in *exact* arithmetic, but the engine's kernels are shape-dependent.** The reference path decodes one token at a time (S=1). MTP verification runs a *batch* forward (S>1) over the draft positions — and colibrì's integer-dot kernels take different code paths by batch shape (the int4 IDOT kernel only engages at S≥2; S=1 uses a different path). Different path = different rounding of the same math. On a **near-tie** between two tokens, that tiny rounding difference flips the greedy argmax — and the token stream forks. Same mechanism for the CUDA tier: the GPU kernel rounds differently from the CPU kernel, so 2/5 near-ties flip there.

So every token MTP emits *is* the argmax of a valid forward of the model — the continuations you saw are all correct — but the batched-verification path is not **bit-identical** to the S=1 path. The word 'lossless' in the README is only true in exact arithmetic; in practice, with quantized shape-dependent kernels, it should say 'produces a valid greedy continuation, but not necessarily byte-identical to non-speculative greedy.' **I'm correcting the README** — thank you for holding it to its word.

**Two things that would confirm the mechanism (and help me):**
1. `IDOT=0 MTP=1` (exact f32 kernels, no shape dependence) vs your `MTP=0` reference. My prediction: the divergences shrink dramatically or vanish — because IDOT=0 removes the S=1-vs-S≥2 rounding gap. If they vanish, that pins it 100% on the quantized kernel path.
2. The `for for` duplicate (prompt 1) is the one I want to be sure about — a near-tie fork can produce a slightly-worse-but-still-valid continuation, which is most likely what it is, but if `IDOT=0` still shows a duplicate there, share the exact prompt and I'll dig into the verification indexing directly.

For strict reproducibility today: `DRAFT=0` (or the adaptive guard) gives byte-exact greedy. MTP trades that last bit of bit-identity for speed — which is the right default on warm caches, but the README shouldn't have claimed both. Fixing it now.

**Comment by dnnspaul** [comment]
Ran your experiment. **Your theory holds — 5/5. And the `for for` duplicate does *not* survive exact kernels, so there is no indexing bug to chase.**

4 configs × 5 prompts, CPU only, greedy, `--ngen 32`, `RAM_GB=65`. OMP hot-thread tuning left **on** this time (I had wrongly disabled it in my #101 numbers — corrected there).

| | `IDOT=1 MTP=0` (A) | `IDOT=0 MTP=0` (B) | `IDOT=0 MTP=1` (C) | `IDOT=1 MTP=1` (D) |
|---|---|---|---|---|
| **B == C (MTP lossless with f32 kernels)** | — | — | **5/5 ✅** | — |
| A != D (originally reported divergence) | — | — | — | 3/5 |
| A != B (IDOT alone changes output) | — | **3/5** | — | — |

### 1. `IDOT=0 MTP=1` is byte-identical to `IDOT=0 MTP=0` on **5/5 prompts**

With the shape-dependent rounding gap removed, **MTP becomes exactly lossless** — including on the two prompts where it diverged with quantized kernels. That is your prediction, confirmed. MTP's verification/acceptance logic is sound; the fork was never a logic bug.

### 2. The `for for` duplicate disappears under `IDOT=0` — no indexing bug

You asked me to send the prompt if the duplicate survived exact kernels. **It doesn't.** It appears *only* in `IDOT=1 MTP=1`:

```
A (idot=1, mtp=0): ...renowned for its rich and complex history. One of its most famous landmarks is the Brandenburg Gate
D (idot=1, mtp=1): ...renowned for for its rich history and vibrant modern energy...   <- duplicate
C (idot=0, mtp=1): The capital of Germany is Berlin. One of its most famous landmarks is the Brand enburg Gate (Brandenburger Tor)...
B (idot=0, mtp=0): (identical to C)
```

So the duplicate is a *near-tie argmax flip under quantized-kernel rounding*, exactly as you said — not double-committed bookkeeping in the verification path. Nothing to dig into.

### 3. The bit that surprised me: `IDOT` **alone** changes greedy output on 3/5 prompts

Comparing A vs B — **no speculation at all in either**, only the kernel family swapped — the output forks on 3 of 5 prompts. So the shape-dependent quantized kernels move the argmax *on their own*; MTP was never the cause, it just exposed the same rounding through its batched (S≥2) forward. That strengthens your explanation: the root cause is quantized-kernel rounding sensitivity, and MTP/CUDA/batched-prefill are all just different ways to hit it.

### 4. …which suggests int4 GLM-5.2 is sitting very close to argmax ties, and that can degrade text

The two degenerate outputs I've now seen came from *different* paths:

- `IDOT=1 MTP=1` → `renowned for **for** its rich history`
- `IDOT=0` (the numerically **most exact** path!) → `the **Brand enburg** Gate`

Neither path is "broken." The model is quantized closely enough to ties that *any* rounding change can flip a token, and some of those flips land on degenerate text. Note the exact-f32 path produced its own artifact — so this is not "quantized kernels are wrong," it's "int4 leaves very little margin."

That raises the stakes on the quality question you flagged as most valuable in the README. **I'm running `./coli bench` (hellaswag/arc_challenge/mmlu) on this machine now, on the CPU reference path (`IDOT=1 MTP=0 COLI_CUDA=0`), and will open a separate issue with the numbers.** If int4 lands within a few points of published GLM-5.2 scores, these near-tie flips are cosmetic; if it doesn't, they may be a symptom of quantization sitting too tight, and mixed/grouped-scale quantization gets a lot more interesting.

### Suggested README wording

Your proposed phrasing is right, and I'd extend it slightly, since MTP is not the only path that forks:

> Speculative decoding (and the CUDA expert tier, and batched prefill) produce a **valid greedy continuation, but not necessarily byte-identical** to single-token greedy: colibrì's quantized kernels are shape-dependent, so a batched or GPU forward rounds differently and can flip near-ties. For byte-exact reproducibility use `DRAFT=0` (and `IDOT=0` if you also want kernel-family independence).

Happy to run anything else. `IDOT=0` costs surprisingly little on this box, so "exact mode" is a viable setting here if you ever want a reproducibility flag that bundles `DRAFT=0 IDOT=0 COLI_CUDA=0`.

**Comment by JustVugg** [comment]
This is definitive — thank you. You confirmed the mechanism cold and then found something more important than the original bug.

The three confirmations, exactly as hoped:
1. **`IDOT=0 MTP=1` == `IDOT=0 MTP=0` on 5/5** — MTP is *exactly* lossless once the shape-dependent rounding gap is removed. The verification/acceptance logic was never the issue.
2. **The `for for` duplicate vanishes under `IDOT=0`** — a near-tie argmax flip under quantized-kernel rounding, not double-committed bookkeeping. No indexing bug to chase, as you proved rather than assumed.
3. **`IDOT` alone forks 3/5 with no speculation at all** — this is the real headline. Swapping only the kernel family (int4 IDOT vs f32) moves the greedy argmax on its own, so MTP/CUDA/batched-prefill aren't *causes*, they're just three different ways to trip the same rounding sensitivity. That reframes the whole thing correctly.

And your point 4 is the one that matters most: **int4 GLM-5.2 sits very close to argmax ties — so little margin that any rounding change can flip a token, and some flips land on degenerate text** (even the exact-f32 path produced its own `Brand enburg` artifact). That's not 'kernels are wrong,' it's 'int4 leaves almost no headroom,' and it's a *quantization-quality* statement, not a correctness one. It's exactly why better quantization is the highest-value lever on the roadmap — grouped-scale / error-feedback (GPTQ-family) quantization exists precisely to widen that margin, and on real GLM-5.2 experts we've measured it cut output error ~5× vs round-to-nearest at equal bits. More margin = fewer near-ties = fewer degenerate flips *and* smaller files. Your observation is the qualitative side of that quantitative case.

**Your `./coli bench` run is the single most valuable thing anyone can contribute right now** — the first community quality measurement of the int4 model. If int4 lands within a few points of published GLM-5.2 scores, these flips are cosmetic; if it trails, it's evidence the quantization is too tight and grouped/mixed-scale is worth the effort. Please do open that separate issue with the numbers — I'll be watching for it.

Taking your extended README wording (it's better than mine — it covers CUDA and batched prefill, not just MTP), and the **`exact mode` bundle (`DRAFT=0 IDOT=0 COLI_CUDA=0`) as a single reproducibility flag** is a genuinely good idea — noting it as a small feature to add. Since `IDOT=0` costs little on your box, a documented 'bit-exact reproducible' mode is worth having. Superb work, start to finish.

**Comment by JustVugg** [comment]
Resolved: the mechanism is fully confirmed (shape-dependent int-kernel rounding, not a verification bug — MTP is lossless once IDOT rounding matches), and it's documented in the README with the exact-mode recipe `DRAFT=0 IDOT=0 COLI_CUDA=0` for byte-identical greedy. Nothing left open here — closing. Thanks for the definitive cold-repro.
